Interactive object tracing

rcooke-ast opened the pull request:
A simple utility to interactively add/delete/modify object traces in PypeIt. This script is designed to be run during the data reduction (i.e. not a stand-alone script). There are several trace methods that are implemented, as you can see from the screenshot below:

![image](https://user-images.githubusercontent.com/12585758/67809146-1a00b780-fa90-11e9-9168-fcba61e948c4.png)

At this stage, the interactive object tracing GUI will only be launched if there are no objects detected, or if the user sets interacting tracing to be true using the command:

```
[scienceimage]
  interactive_objfind = True
```

The `manual` option allows users to basically put down an object trace wherever they like. This GUI also allows the user to update the FWHM using the small object profile window. Future improvements might include:
* Changing the image colour scale with a RMB click and drag
* Add your thoughts here...

One might also consider making a similar utility for the slit tracing...

profxj commented on pypeit/core/gui/object_find.py:
```diff
@@ -0,0 +1,684 @@
+import os
```
Please add a line or two describing this module.

rcooke-ast replied:
OK done!

profxj commented on pypeit/core/gui/object_find.py:
```diff
@@ -0,0 +1,684 @@
+            frame (ndarray): Sky subtracted science image
+            trace_dict (dict): Dictionary containing slit and object trace information
+            sobjs (SpecObjs): SpecObjs Class
+            slit_ids (list, None): List of slit ID numbers
```
nit:  I think it is supposed to be  `list, optional`

rcooke-ast replied:
Thanks!

profxj commented on pypeit/core/gui/object_find.py:
```diff
@@ -0,0 +1,684 @@
+            slit_ids (list, None): List of slit ID numbers
+
+        Returns:
+            ofgui (ObjFindGUI): Returns an instance of the ObjFindGUI class
```
nit:  Kyle prefers we do:

```
ObjFindGui:  desc
```

i.e. no naming the variable

rcooke-ast replied:
OK - thanks for pointing that out - I've fixed all instances of this in the PR!

profxj commented:
A couple of extra points:

1.  Any thoughts on how to test this?  At the least, might you be able to 
add a separate script to the Dev Suite that lets us run it on a small dataset?

2.  For the auto-launch option (i.e. when no object is detected), I presume that
is on a per-slit basis.

3.  Would be great if you generated an `.rst` file in docs/ with some basic instructions.

rcooke-ast commented:
OK - I've just submitted a PR in the dev suite to give folks an idea of how to use this utility. I've also just added some detailed docs on how to interrogate the GUI, just in case anyone wishes to give it a test run.

jhennawi commented on pypeit/specobjs.py:
```diff
@@ -608,6 +608,12 @@ def __getattr__(self, k):
         # Recast as an array
         return lst_to_array(lst)
 
+    def __getstate__(self):
```
I'm not following why these changes were needed?

rcooke-ast replied:
I needed to save the SpecObjs into a pickle file to be stored in the dev suite, and without these being set, I ran into a recursion problem.

jhennawi replied:
I don't think we need this function anymore, since now interactive will just create manual parameters right? Maybe it does not hurt to have it though, but ask @profxj who has reworked specobj.

rcooke-ast replied:
Agreed - removed.

jhennawi commented:
@rcooke-ast I tried to take a look at this, but the code crashes for all instruments except WHT_ISIS. You introduced a bug in PR #775. Please solve the bpm issue first so that I can test this functionality on dev suite spectrographs, and I'll give it a spin and review this PR.

rcooke-ast commented:
@jhennawi - I'll fix the BPM generation when I submit my next PR for the red arm of WHT/ISIS. I've merged your hotfix into this PR - thanks!

jhennawi commented on pypeit/core/gui/object_find.py:
```diff
@@ -0,0 +1,686 @@
+        self.replot()
+
+    def make_objprofile(self):
+        """Generate an object profile from the traces
```
So if the WHT/ISIS detector does not include the slit boundaries because the entire detector is illuminated, then you need to use the standard as the crutch for slit tracing. The standard already gets passed to the object finding code to be used as a crutch, so I think this should be dealt with correctly, but it is worth checking.

jhennawi commented on doc/object_finding.rst:
```diff
@@ -90,4 +90,93 @@ xedge          Any         float; default=0.03      Ignore any objects within xe
+above, hover the mouse to the location you would like to lay down an
+object trace, and press the 'a' key on the keyboard.
+
+When using the "manual" object trace method, you need to define the
```
Yes see my comment above. The solution here is to require the standard, and to use it to create the samshed flux image and to use the standard everywhere in place of the slit boundaries. If you don't have a standard taken the same night, any standard with that setup would do.

jhennawi commented:
One last comment @rcooke-ast. I fear that there are going to be conflicts with your branch and the edge_merge_trace PR, as that has made major changes to the extract.py objfind. You should discuss the merge with @kbwestfall. Ideally we would have had you branch off of the edge_merge_tracebranch.

profxj commented:
One additional set of comments from me, as I had a chance to touch the GUI on the airplane yesterday and took these (not so easy to understand) notes:

1.  Provide feedback when LMB is pressed
2.  Note in Help that panning requires clicking on Navigate thingy first.  Or it did for me.
3.  RMB behavior is funny.  I'd stick to the matplotlib magnifying glass
4.  Is there a way to have it center on the cursor when a key is pressed, e.g. "c"?
5.  It isn't clear what to do when I am done, nor to really know I am done.
6.  How is the object profile being generated?
7.  What is the vertical, red dashed line?

rcooke-ast commented:
@profxj - I'm surely being daft, but I can't see any additional comments from you. Please let me know if I'm missing something!

@jhennawi - I don't think the conflicts will be too severe. Afterall, this is designed to be run as a standalone GUI now. In principle, we don't want folks to run this at runtime. I can see the clear advantages for avoiding that. However, I've kept the ability to use the GUI in runtime if the user *really* wants to!

rcooke-ast commented:
Hi @profxj - here are some responses to your queries:

> 1. Provide feedback when LMB is pressed

I've added feedback when an object is selected, when the FWHM is updated, and when a different method is selected.

> 2. Note in Help that panning requires clicking on Navigate thingy first.  Or it did for me.

I've added this to the help menu

> 3. RMB behavior is funny.  I'd stick to the matplotlib magnifying glass

OK - both functionalities are available. I'm not sure how to change or turn off the RMB zooming behaviour, but this doens't stop a user from using the magnifying glass.

> 4. Is there a way to have it center on the cursor when a key is pressed, e.g. "c"?

Done - the new keyword to do this is the letter "c"

> 5. It isn't clear what to do when I am done, nor to really know I am done.

The docs say that you need to click the exit button, but I've now added a note to the help menu.

> 6. How is the object profile being generated?

It's stored in a dictionary, and is based on the information generated from the data reduction.

> 7. What is the vertical, red dashed line?

That's a bad object trace (in fact, it's poorly defined bad pixels, which PypeIt picks up as an object trace). This is an example of an object trace that should be deleted. I've added a description of the different coloured lines to the help menu.

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -22,6 +22,7 @@
 from pypeit.core import pydl
 from pypeit.core import pixels
 from pypeit.core import arc
+from pypeit.core.gui import object_find as gui_object_find
```
This import dependency should be removed.

rcooke-ast replied:
done

jhennawi commented on pypeit/core/extract.py:
```diff
@@ -1504,12 +1506,9 @@ def objfind(image, thismask, slit_left, slit_righ, inmask=None, fwhm=3.0, maxdev
-    ## Okay now loop over all the regular aps and exclude any which within the fwhm of the hand_extract_APERTURES
-    if nobj_reg > 0 and hand_extract_dict is not None:
+    # Okay now loop over all the regular aps and exclude any which within the fwhm of the hand_extract_APERTURES
+    # We're going to assume that if interactive updates were made that the user has manually assigned the traces.
```
Can we remove this comment about interactive updates, since it is a bit confusing given that the interactive part is removed now.

rcooke-ast replied:
Good catch - yes I've removed line 1510

jhennawi commented on pypeit/reduce.py:
```diff
@@ -555,6 +553,9 @@ def find_objects_pypeline(self, image, std=False, ir_redux=False, std_trace = No
         SHOW_TRACE:  bool
           Generate QA  showing traces identified. Requires an open ginga RC modules window
 
+        INTERACTIVE:  bool
```
Can we remove this from the docs since the interactive feature is not there anymore.

rcooke-ast replied:
I thought I caught them all - thanks for catching that!

jhennawi commented:
Hi Ryan,

Can you please test on a few fast dev suite setups, Shane Kast, and also one multislit setup to make sure all is kosher. I don't think the whole dev suite is required here, but you should also run python setup.py test since that is fast.

rcooke-ast commented:
Results from running Shane Kast:

```
Finished running pypeit on shane_kast_blue_600_4310_d55 --- PASSED
[INFO]    :: pypeitsetup.py 470 run() - Inspect the .sorted file
[INFO]    :: util.py 606 make_pypeit_file() - PypeIt file written to: /Users/rcooke/Software/PypeIt-development-suite/REDUX_OUT/Shane_Kast_blue/600_4310_d55/shane_kast_blue_A/shane_kast_blue_A.pypeit


Directory: /Users/rcooke/Software/PypeIt-development-suite/REDUX_OUT/Shane_Kast_blue/600_4310_d55/shane_kast_blue_A
Command line: run_pypeit shane_kast_blue_A.pypeit -o
Wrote: QA/MF_A.html
Running pypeit on shane_kast_blue_A.pypeit --- PASSED
[INFO]    :: run_pypeit.py 94 main() - Data reduction complete
[INFO]    :: run_pypeit.py 96 main() - Generating QA HTML


Directory: /Users/rcooke/Software/PypeIt-development-suite/REDUX_OUT/Shane_Kast_blue/600_4310_d55/shane_kast_blue_A/Science
Command line: pypeit_flux_spec /Users/rcooke/Software/PypeIt-development-suite/fluxing_files/shane_kast_blue.flux
Writing the parameters to fluxing.par
Running pypeit_flux on shane_kast_blue_A.pypeit --- FLUXING PASSED
[INFO]    :: flux_calib.py 365 extinction_correction() - Extinction data covered the whole spectra. Correct it!
[INFO]    :: specobj.py 440 apply_flux_calib() -  Masking bad pixels
[INFO]    :: specobjs.py 409 write_to_fits() - Wrote 1D spectra to J1217+3905_2.fits


Directories:
         Raw data: /Users/rcooke/Software/PypeIt-development-suite/RAW_DATA/Shane_Kast_blue/452_3306_d57
    PypeIt output: /Users/rcooke/Software/PypeIt-development-suite/REDUX_OUT/Shane_Kast_blue/452_3306_d57

Directory: /Users/rcooke/Software/PypeIt-development-suite/REDUX_OUT/Shane_Kast_blue/452_3306_d57
Command line: run_pypeit shane_kast_blue_452_3306_d57.pypeit -o
Wrote: QA/MF_A.html
Running pypeit on shane_kast_blue_452_3306_d57.pypeit --- PASSED
[INFO]    :: run_pypeit.py 94 main() - Data reduction complete
[INFO]    :: run_pypeit.py 96 main() - Generating QA HTML

--- PYPEIT DEVELOPMENT SUITE PASSED 4/4 TESTS ---
```
